File: src/main/java/wraith/fabricaeexnihilo/api/registry/CrucibleHeatRegistry.java
package wraith.fabricaeexnihilo.api.registry;

import net.minecraft.block.Block;
import net.minecraft.fluid.Fluid;
import net.minecraft.item.BlockItem;
import net.minecraft.tag.Tag;
import wraith.fabricaeexnihilo.api.crafting.BlockIngredient;
import wraith.fabricaeexnihilo.api.crafting.FluidIngredient;
import wraith.fabricaeexnihilo.api.recipes.crucible.CrucibleHeatRecipe;

import java.util.Collection;

public interface CrucibleHeatRegistry extends Registry<CrucibleHeatRecipe> {

    default boolean register(BlockIngredient blocks, FluidIngredient fluids, int heat) {
        return register(new CrucibleHeatRecipe(blocks, fluids, heat));
    }

    default boolean register(Tag.Identified<Fluid> tag, int heat) {
        return register(BlockIngredient.EMPTY, new FluidIngredient(tag), heat);
    }

    default boolean register(Block block, int heat) {
        return register(new BlockIngredient(block), FluidIngredient.EMPTY, heat);
    }

    default boolean register(Fluid fluid, int heat) {
        return register(BlockIngredient.EMPTY, new FluidIngredient(fluid), heat);
    }

    int getHeat(Block block);

    int getHeat(Fluid fluid);

    int getHeat(BlockItem item);

    // All recipes, chunked/broken up for pagination
    Collection<CrucibleHeatRecipe> getREIRecipes();

}

File: src/main/kotlin/exnihilofabrico/registry/CrucibleHeatRegistry.kt
package exnihilofabrico.registry

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import exnihilofabrico.ExNihiloFabrico
import exnihilofabrico.api.crafting.FluidIngredient
import exnihilofabrico.api.recipes.CrucibleHeatRecipe
import exnihilofabrico.api.registry.ICrucibleHeatRegistry
import exnihilofabrico.compatibility.modules.MetaModule
import net.minecraft.block.Block
import net.minecraft.fluid.Fluid
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.recipe.Ingredient
import java.io.File
import java.io.FileReader
import java.io.FileWriter

data class CrucibleHeatRegistry(val registry: MutableList<CrucibleHeatRecipe> = ArrayList()): AbstractRegistry<MutableList<CrucibleHeatRecipe>>(), ICrucibleHeatRegistry {

    override fun clear() = registry.clear()

    override fun register(recipe: CrucibleHeatRecipe)  = registry.add(recipe)
    override fun register(block: Ingredient, heat: Int) = registry.add(CrucibleHeatRecipe(block, null, heat))
    override fun register(fluid: FluidIngredient, heat: Int) = registry.add(CrucibleHeatRecipe(null, fluid, heat))

    override fun register(block: Block, heat: Int) = register(Ingredient.ofItems(block), heat)
    override fun register(fluid: Fluid, heat: Int) = register(FluidIngredient(fluid), heat)

    override fun getHeat(block: Block) = registry.firstOrNull { it.test(block) }?.value ?: 0
    override fun getHeat(fluid: Fluid) = registry.firstOrNull { it.test(fluid) }?.value ?: 0
    override fun getHeat(item: Item) = registry.firstOrNull { it.test(item) }?.value ?: 0
    override fun getHeat(stack: ItemStack) = registry.firstOrNull { it.test(stack) }?.value ?: 0

    override fun serializable() = registry
    override fun registerJson(file: File) {
        if(file.exists()){
            val json: MutableList<CrucibleHeatRecipe> = gson.fromJson(FileReader(file), SERIALIZATION_TYPE)
            json.forEach { register(it) }
        }
    }

    companion object {
        val SERIALIZATION_TYPE = object: TypeToken<MutableList<CrucibleHeatRecipe>>(){}.type
        fun fromJson(file: File) = fromJson(file, {CrucibleHeatRegistry()}, MetaModule::registerCrucibleHeat)
    }
}